From c060213236a27196b42ae6a8150a24135ee99992 Mon Sep 17 00:00:00 2001
From: timyao <tim.yao@amlogic.com>
Date: Thu, 19 Dec 2013 00:57:32 -0800
Subject: [PATCH 2643/5965] PD #83025. Change the method to solve top level DOS
 register access HW bug. Add protocol in hcodec ucode and make the work around
 only trigger once when 4k2k decoder is running.

---
 drivers/amlogic/amports/encoder.c             |  73 +++-
 .../m8/ucode/encoder/full_encoder_linux.h     | 324 +++++++++---------
 drivers/amlogic/amports/vdec.c                |  39 ---
 drivers/amlogic/amports/vdec.h                |   1 -
 drivers/amlogic/amports/vh264_4k2k.c          |   4 +
 drivers/amlogic/amports/vreal.c               |   4 -
 6 files changed, 237 insertions(+), 208 deletions(-)

diff --git a/drivers/amlogic/amports/encoder.c b/drivers/amlogic/amports/encoder.c
index 851600caa7e1..ddb3835684ca 100755
--- a/drivers/amlogic/amports/encoder.c
+++ b/drivers/amlogic/amports/encoder.c
@@ -16,6 +16,7 @@
 #include <linux/slab.h>
 #include <linux/dma-mapping.h>
 #include <linux/platform_device.h>
+#include <linux/spinlock.h>
 #include <mach/am_regs.h>
 #include <mach/power_gate.h>
 #include <plat/io.h>
@@ -26,7 +27,9 @@
 #include <linux/amlogic/amports/vframe.h>
 #include <linux/amlogic/amports/vframe_provider.h>
 #include <linux/amlogic/amports/vframe_receiver.h>
+#include <linux/amlogic/amports/vformat.h>
 #include "vdec_reg.h"
+#include "vdec.h"
 #include <linux/delay.h>
 #include <linux/poll.h>
 #include <linux/of.h>
@@ -126,6 +129,8 @@ static wait_queue_head_t avc_wait;
 atomic_t avc_ready = ATOMIC_INIT(0);
 static struct tasklet_struct encode_tasklet;
 
+static DEFINE_SPINLOCK(lock);
+
 static const char avc_dec_id[] = "avc-dev";
 
 #define AMVENC_BUFFER_LEVEL_480P   0
@@ -940,12 +945,11 @@ static void avc_prot_init(void)
     data32 = data32 | (1<<0); // set pop_coeff_even_all_zero
     WRITE_HREG(VLC_CONFIG , data32);	
     
-        /* clear mailbox interrupt */
+    /* clear mailbox interrupt */
     WRITE_HREG(HCODEC_ASSIST_MBOX2_CLR_REG, 1);
 
     /* enable mailbox interrupt */
     WRITE_HREG(HCODEC_ASSIST_MBOX2_MASK, 1);
-    
 }
 
 void amvenc_reset(void)
@@ -1042,6 +1046,56 @@ s32 amvenc_loadmc(const u32 *p)
 
     return ret;
 }
+
+#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
+const u32 fix_mc[] __attribute__ ((aligned (8))) = {
+    0x0809c05a, 0x06696000, 0x0c780000, 0x00000000
+};
+
+
+/*
+ * DOS top level register access fix.
+ * When hcodec is running, a protocol register HCODEC_CCPU_INTR_MSK
+ * is set to make hcodec access one CBUS out of DOS domain once
+ * to work around a HW bug for 4k2k dual decoder implementation.
+ * If hcodec is not running, then a ucode is loaded and executed
+ * instead.
+ */
+void amvenc_dos_top_reg_fix(void)
+{
+    bool hcodec_on;
+    unsigned long flags;
+
+    spin_lock_irqsave(&lock, flags);
+
+    hcodec_on = vdec_on(VDEC_HCODEC);
+
+    if ((hcodec_on) && (READ_VREG(HCODEC_MPSR) & 1)) {
+        WRITE_HREG(HCODEC_CCPU_INTR_MSK, 1);
+        spin_unlock_irqrestore(&lock, flags);
+        return;
+    }
+
+    if (!hcodec_on) {
+        vdec_poweron(VDEC_HCODEC);
+    }
+
+    amhcodec_loadmc(fix_mc);
+
+    amhcodec_start();
+
+    udelay(1000);
+
+    amhcodec_stop();
+
+    if (!hcodec_on) {
+        vdec_poweroff(VDEC_HCODEC);
+    }
+
+    spin_unlock_irqrestore(&lock, flags);
+}
+#endif
+
 #if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON6TV
 #define  DMC_SEC_PORT8_RANGE0  0x840
 #define  DMC_SEC_CTRL  0x829
@@ -1057,12 +1111,16 @@ void enable_hcoder_ddr_access(void)
 
 static s32 avc_poweron(void)
 {
+	unsigned long flags;
 	u32 data32 = 0;
 	data32 = 0;
 	enable_hcoder_ddr_access();
+
 #if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
 	CLK_GATE_ON(DOS);
 
+	spin_lock_irqsave(&lock, flags);
+
 	data32 = READ_AOREG(AO_RTI_PWR_CNTL_REG0);
 	data32 = data32 & (~(0x18));
 	WRITE_AOREG(AO_RTI_PWR_CNTL_REG0, data32);
@@ -1096,14 +1154,22 @@ static s32 avc_poweron(void)
 	data32 = READ_VREG(DOS_GEN_CTRL0);
 	data32 = data32 & 0xFFFFFFFE;
 	WRITE_VREG(DOS_GEN_CTRL0, data32);
+
+	spin_unlock_irqrestore(&lock, flags);
 #endif
+
 	mdelay(10);
+
 	return 0;
 }
 
 static s32 avc_poweroff(void)
 {
 #if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
+	unsigned long flags;
+
+	spin_lock_irqsave(&lock, flags);
+
 	// enable HCODEC isolation
 	WRITE_AOREG(AO_RTI_GEN_PWR_ISO0, READ_AOREG(AO_RTI_GEN_PWR_ISO0) | 0x30);
 	// power off HCODEC memories
@@ -1112,6 +1178,9 @@ static s32 avc_poweroff(void)
 	hvdec_clock_disable();
 	// HCODEC power off
 	WRITE_AOREG(AO_RTI_GEN_PWR_SLEEP0, READ_AOREG(AO_RTI_GEN_PWR_SLEEP0) | 0x3);
+
+	spin_unlock_irqrestore(&lock, flags);
+
 	// release DOS clk81 clock gating
 	CLK_GATE_OFF(DOS);
 #else
diff --git a/drivers/amlogic/amports/m8/ucode/encoder/full_encoder_linux.h b/drivers/amlogic/amports/m8/ucode/encoder/full_encoder_linux.h
index 2914925e67c1..dfb00debe447 100755
--- a/drivers/amlogic/amports/m8/ucode/encoder/full_encoder_linux.h
+++ b/drivers/amlogic/amports/m8/ucode/encoder/full_encoder_linux.h
@@ -1,175 +1,175 @@
 const u32 MicroCode[] __attribute__((aligned(8))) = {
-	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc013f40,0x0000000,
-	0xc000e80,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0x69fffe2,0x6c01c22,
-	0xc780e40,0x0000000,0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x6810001,0x6800000,0xd000001,0x7400040,0xc0009c0,0x0000000,0xc013ec0,0x0000000,
+	0xc000e80,0x0000000,0x6bffc40,0x7c00000,0x6030400,0x0400000,0xc780ec0,0x0000000,
+	0xc780000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
-	0x0000000,0xc792940,0x0000000,0xc78ccc0,0x64d3008,0xc780900,0x0000000,0xc792980,
+	0x0000000,0xc7928c0,0x0000000,0xc78cc40,0x64d3008,0xc780900,0x0000000,0xc792900,
 	0x0000000,0xcc00000,0x0400000,0x80d1a00,0x80ac300,0x80ac400,0xcc00000,0x0000000,
 	0x64ac808,0x7c00108,0x64ac908,0x7c00148,0x64ac608,0x7c00188,0x64ac508,0x7c001c8,
 	0x64ac708,0x7c00208,0x64aca08,0x7c00388,0x64acb08,0x7c00048,0x64ace08,0x7c003c8,
 	0xcc00000,0x0000000,0xcc00000,0x0000000,0x80d2001,0x64d2008,0xbefc048,0xcc00000,
-	0x0000000,0xa614022,0x0000000,0x6696022,0x69fffe2,0x6c01c22,0x44018a2,0x64ac021,
-	0xaae4021,0xaa34061,0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,0xc784dc0,0x0000000,
-	0xa6100e1,0x0000000,0xc782980,0x0000000,0xc7ffb40,0x0000000,0xc07f480,0x0000000,
-	0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,
-	0x6800048,0x6c00408,0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,
-	0x6a10808,0x60d1f08,0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,0xc07f380,0x80d2100,
-	0x7800148,0x4404208,0xc07f280,0x60d2108,0xc07f200,0x80d2100,0x7800108,0x4404208,
-	0xc07f100,0x60d2108,0xc07f080,0x80d2101,0x80d1f20,0x64d3109,0x9010209,0x400f208,
-	0x5804208,0x4401208,0x7c00308,0xc07ee40,0x60d2108,0x9210209,0x400f208,0x5804208,
-	0x4401208,0x7c00348,0xc07ec80,0x60d2108,0x80d1f30,0x80d1f30,0x64d3108,0x9210208,
-	0x6810e09,0xa803248,0xc7802c0,0x80d1f20,0x80d1f30,0xc07e9c0,0x80d2100,0xc07e940,
-	0x80d2100,0xc07e8c0,0x80d2100,0xc07e840,0x80d2104,0x80d1f20,0x80d1f30,0x6bc0008,
-	0x60d1f08,0x0000000,0x80ac007,0x8007801,0xc7fe740,0x0000000,0xc07e080,0x0000000,
-	0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,
-	0x6800048,0x6c00408,0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,
-	0xc07e100,0x80d2100,0xc07e080,0x80d2100,0x80d1f20,0x80d1f20,0xc07df80,0x80d2100,
-	0xc07df00,0x80d2100,0xc07de80,0x80d2100,0x80d1f20,0x80d1f40,0x7800048,0xc007d40,
-	0x441a208,0xc07dcc0,0x60d2108,0xc007c40,0x6800008,0xc07dbc0,0x60d2108,0xc07db40,
-	0x80d2100,0x80d1f20,0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,
-	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,
-	0x0000000,0x6800008,0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,
-	0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,
-	0x60d1b08,0x64d1608,0xbefc108,0x0000000,0x80ac008,0x8007801,0xc7fd2c0,0x0000000,
-	0xc010400,0x0000000,0x6490b08,0x9c08608,0x9c08648,0x6090b08,0x6401108,0x9c08748,
-	0x9c08788,0x6001108,0x6401008,0x9808008,0x6001008,0x9c08008,0x6001008,0xc010840,
-	0x0000000,0xc00ff40,0x0000000,0xc07c740,0x0000000,0x64ac408,0x4001208,0x60ac408,
-	0x64f4908,0x9c08048,0x60f4908,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,
-	0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a19408,0x60d1f08,
-	0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9008248,0x7c00089,0x9108248,0x7c000c9,
-	0x9210208,0xc07c4c0,0x60d2108,0xc07c440,0x80d2107,0x64d3109,0x9010209,0x400f208,
-	0x5804208,0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,
-	0x6800008,0x7c00248,0x60f1b08,0x7800309,0x946d209,0x60d3208,0x6940008,0x6c00008,
-	0x2008308,0x60f2008,0x6803908,0x64d3309,0x9008249,0x9508209,0x9808508,0x9808548,
-	0x98087c8,0x60f1f08,0x7800309,0x7800348,0x960f248,0x60f1e09,0xc07bb80,0x80d2100,
-	0x7800148,0x7800189,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x78001c8,
-	0xc07b900,0x60d2108,0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,
-	0x60d1f09,0x80d1f20,0x80d1f20,0x64f1c08,0x9206208,0x7800049,0xc005580,0x2408248,
-	0xc07b500,0x60d2108,0x80d3d00,0x64acd08,0x60f4208,0xc0055c0,0x0000000,0x680d703,
-	0x68133c4,0x80d300f,0x64d0008,0x9808008,0x9808048,0x9808088,0x9c080c8,0x9c08148,
-	0x9c08188,0x9c081c8,0x60d0008,0x64acf08,0x60f4808,0x6800289,0x78003c8,0x9485248,
-	0x60f4009,0x80ac005,0x8007801,0xc7faf80,0x0000000,0xc00eac0,0x0000000,0x64f4008,
-	0x9c08048,0x60f4008,0xc00df80,0x0000000,0x6490b08,0x9c08608,0x9808648,0x6090b08,
-	0x6401108,0x9808748,0x9c08788,0x6001108,0x6401008,0x9808008,0x6001008,0x9c08008,
-	0x6001008,0xc00e3c0,0x0000000,0xc00dac0,0x0000000,0xc07a2c0,0x0000000,0x64f4308,
-	0x9c08048,0x9c080c8,0x60f4308,0x64f1508,0x9c08508,0x9c084c8,0x9c08748,0x9c08448,
-	0x9c08488,0x9c08348,0x9c08308,0x9808248,0x9808008,0x60f1508,0x80d3400,0x9808488,
-	0x9808448,0x9808348,0x9808308,0x98085c8,0x9808608,0x9808648,0x9c08688,0x98086c8,
-	0x9c08248,0x9c08008,0x60f1508,0x64f4908,0x9c08048,0x9c081c8,0x9808008,0x60f4908,
-	0x78003c9,0x5805249,0x96e1209,0x5801249,0x9761209,0x98087c8,0x9808788,0x9808748,
-	0x9808688,0x9808648,0x9808608,0x9808288,0x9c08248,0x98081c8,0x9808088,0x9808048,
-	0x60f4908,0x4801249,0xaa10049,0x0000000,0xc781640,0x0000000,0x6800c08,0x6800809,
-	0x94e8209,0x6800c09,0x9608209,0x6801009,0x9708209,0x60f5608,0x6800008,0x6800009,
-	0x9508209,0x60f5708,0x680c0c8,0x6c00008,0x60f5808,0x68000c8,0x68000c9,0x9484209,
-	0x68000c9,0x9504209,0x6800049,0x9581209,0x6800009,0x95a1209,0x6800049,0x95c5209,
-	0x60f5908,0x6800388,0x6800489,0x9508209,0x6800149,0x9604209,0x68000c9,0x9684209,
-	0x6800109,0x9704209,0x6800089,0x9704209,0x60f5a08,0x6800008,0x6800009,0x9508209,
-	0x6800809,0x9608209,0x6801409,0x9708209,0x60f5b08,0x6801008,0x6800009,0x9508209,
-	0x6801809,0x9608209,0x6803009,0x9708209,0x60f5c08,0x6800608,0x6801409,0x9508209,
-	0x6800809,0x9608209,0x6801809,0x9708209,0x60f5d08,0x68000c8,0x6801809,0x948c209,
-	0x6800249,0x9607209,0x6800009,0x96e1209,0x6800c09,0x9708209,0x60f5e08,0x68000c8,
-	0x6800009,0x9461209,0x6800049,0x9481209,0x6800009,0x94a2209,0x6803009,0x950c209,
-	0x6801c09,0x968c209,0x60f5f08,0xc7807c0,0x0000000,0x6804008,0x68000c9,0x9586209,
-	0x6800089,0x9646209,0x9808608,0x9808648,0x9808688,0x98086c8,0x9808708,0x9808748,
-	0x9808788,0x98087c8,0x60f5508,0x6800c08,0x6800809,0x94e8209,0x6800c09,0x9608209,
-	0x6801009,0x9708209,0x60f5608,0x6800c08,0x6801009,0x9508209,0x60f5708,0x680c0c8,
-	0x6c0c0c8,0x60f5808,0x64acc09,0x60f4a09,0x64d3109,0x9010209,0x400f208,0x5804208,
-	0x4401208,0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x7800309,
-	0x7800348,0x958c248,0x6499008,0x9708248,0x60f4f09,0x6bc0008,0x60d1f08,0x64d1e08,
-	0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,
-	0x6a10408,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9210208,0xc077340,
-	0x60d2108,0xc0772c0,0x80d2105,0x6800048,0x7c00248,0x60f1b08,0x7800309,0x946d209,
-	0x60d3208,0x7800348,0x960f248,0x60f1e09,0xc077000,0x80d2100,0x7800148,0x7800189,
-	0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x7800108,0x7800209,0x680040a,
-	0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x80d1f20,0x64f1d08,
-	0x9206208,0x7800049,0xc000a80,0x2408248,0xc076a00,0x60d2108,0x80d3d00,0x80d5300,
-	0x64acd08,0x60f4208,0xc001840,0x0000000,0x68101c3,0x6813c44,0x6800008,0x6c00448,
-	0x60d3008,0x64d0008,0x9c08008,0x9c08048,0x9c08088,0x9c08188,0x9c081c8,0x60d0008,
-	0x9808008,0x9808048,0x9808088,0x98080c8,0x9808148,0x9808188,0x98081c8,0x60d0008,
-	0x6800048,0x7800309,0x948c209,0x60f4b08,0x80ac005,0x8007801,0xc7f62c0,0x0000000,
-	0x0800000,0x7800008,0xc7f59c0,0x6030408,0x6800009,0xb005248,0x2409209,0x5401208,
-	0xcc00000,0x4401208,0xcc00000,0x5401209,0x6bfffda,0x6c0001a,0x6800009,0x680001b,
-	0x6c0001b,0x400071b,0x680000b,0x680c00e,0x680100f,0x740039a,0x400138e,0x7400389,
-	0x400138e,0x740039b,0x400138e,0x740038b,0x400138e,0xaee004f,0x44013cf,0x690000e,
-	0x6c0000e,0x200e30e,0x680000f,0x603510e,0x680400d,0x603520d,0x6a0c00d,0x603500d,
-	0x643500d,0x580f34d,0xbef804d,0x0000000,0x40403cf,0xb611e8f,0x680800d,0xc7ffd00,
-	0x200e34e,0x64d330d,0x900834d,0x680000e,0x680100f,0xb8053cd,0x0000000,0x404038e,
-	0xc7fff40,0x40403cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d371a,0x60d391b,0x60d381a,
-	0x60d3a1b,0xcc00000,0x0000000,0x8098002,0x8098000,0x6a0001d,0x6d0001d,0x6a0001e,
-	0x6d0001e,0x6a0001f,0x6d0001f,0x6a00020,0x6d00020,0x690000a,0x680001b,0x6c0001b,
-	0x400071b,0x6800009,0x680c00e,0x680064f,0x740039b,0x400138e,0x7400389,0x400138e,
-	0x740039d,0x400138e,0x740038a,0x400138e,0x740039e,0x400138e,0x740038a,0x400138e,
-	0x740039f,0x400138e,0x740038a,0x400138e,0x74003a0,0x400138e,0x740038a,0x400138e,
-	0xaeb004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x6803e8d,
-	0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40193cf,
-	0xb611e8f,0x6807d0d,0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680064f,
-	0xb8053cd,0x0000000,0x401938e,0xc7fff40,0x40193cf,0x7c0028e,0x44013cf,0x7c002cf,
-	0x60d391b,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,0x60d4a1d,0x60d3a1b,
-	0x60d4e1d,0x60d4f1e,0x60d501f,0x60d5120,0xcc00000,0x0000000,0xaa10008,0x64d370e,
-	0xcb80003,0x64d390f,0xc7f33c0,0x0000000,0x60d371a,0x64d3510,0x908c250,0xa620009,
-	0x60d391b,0x6bfffc9,0x6c00009,0x60d3809,0x6800009,0x6c00009,0x60d3a09,0x9004250,
-	0xaa1c249,0x80d3601,0x80d3603,0x80d3605,0x80d3606,0xc780200,0x80d360f,0x80d3602,
-	0x80d3603,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0x680c10a,0x7800289,
-	0xa814248,0x2409248,0x4401249,0x5402249,0x680c00a,0x200a289,0x740028e,0x400128a,
-	0x581038e,0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400528a,
-	0x78002c9,0xb40d248,0x7800309,0xb40b248,0x700029a,0x400128a,0x700028e,0x961068e,
-	0x400128a,0x700029b,0x400128a,0x700028f,0xc7f2540,0x96106cf,0x908c210,0x7800289,
-	0x2409248,0x5402249,0x680c00a,0x200a289,0x64d0009,0x9384249,0xa6f8009,0x0000000,
-	0x64d3609,0xa6ec009,0x0000000,0x64d370e,0x64d390f,0x740028e,0x400128a,0x581038e,
-	0x740028e,0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x690000e,
-	0x6c0000e,0x200e30e,0x780028f,0x54033cf,0x200e3ce,0x603510e,0x680400d,0x603520d,
-	0x6a0c00d,0x603500d,0x908c210,0x7800309,0xb830248,0x4001208,0x920c210,0x7800349,
-	0xb82c248,0x6800008,0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,
+	0x0000000,0x6432421,0xaa0c021,0x8032400,0x6699321,0x64ac021,0xaaec021,0xaa34061,
+	0xa60c0a1,0xc7816c0,0x0000000,0xa60c121,0xc784dc0,0x0000000,0xa6100e1,0x0000000,
+	0xc782980,0x0000000,0xc7ffbc0,0x0000000,0xc07f500,0x0000000,0x6bc0008,0x60d1f08,
+	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
+	0x60d1f08,0x6a19c08,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0x6a10808,0x60d1f08,
+	0x6a00008,0x60d1f08,0x6a0a008,0x60d1f08,0xc07f400,0x80d2100,0x7800148,0x4404208,
+	0xc07f300,0x60d2108,0xc07f280,0x80d2100,0x7800108,0x4404208,0xc07f180,0x60d2108,
+	0xc07f100,0x80d2101,0x80d1f20,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
+	0x7c00308,0xc07eec0,0x60d2108,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,
+	0xc07ed00,0x60d2108,0x80d1f30,0x80d1f30,0x64d3108,0x9210208,0x6810e09,0xa803248,
+	0xc7802c0,0x80d1f20,0x80d1f30,0xc07ea40,0x80d2100,0xc07e9c0,0x80d2100,0xc07e940,
+	0x80d2100,0xc07e8c0,0x80d2104,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,
+	0x80ac007,0x8007801,0xc7fe7c0,0x0000000,0xc07e100,0x0000000,0x6bc0008,0x60d1f08,
+	0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,
+	0x60d1f08,0x6a1a008,0x60d1f08,0x64d1e08,0x98087c8,0x60d1e08,0xc07e180,0x80d2100,
+	0xc07e100,0x80d2100,0x80d1f20,0x80d1f20,0xc07e000,0x80d2100,0xc07df80,0x80d2100,
+	0xc07df00,0x80d2100,0x80d1f20,0x80d1f40,0x7800048,0xc007d40,0x441a208,0xc07dd40,
+	0x60d2108,0xc007c40,0x6800008,0xc07dc40,0x60d2108,0xc07dbc0,0x80d2100,0x80d1f20,
+	0x80d5500,0x80d1f20,0x80d1f20,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,
 	0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,
 	0x6e00008,0x60d1c08,0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0x0000000,0x64d1608,0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,
-	0xbefc108,0x0000000,0x7800188,0x4001208,0x7c00188,0x7800208,0x4002208,0x7c00208,
-	0x80ac009,0x8007801,0xc7f0fc0,0x0000000,0x7c00288,0x403f248,0x7c002c9,0x5403208,
-	0x690000e,0x6c0000e,0x200e30e,0x200e20e,0x680400f,0x6b0c010,0x643500d,0x580f34d,
-	0xbef804d,0x0000000,0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,
-	0x6035010,0x643500d,0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,0x680c00a,0x64d4a11,
-	0x64d4b12,0x64d4c13,0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x64d3510,
-	0x908c250,0xa62c009,0x60d391b,0x6a00009,0x6d00009,0x60d4e09,0x60d4f09,0x60d5009,
-	0x60d5109,0x6800009,0x6c00009,0x60d3a09,0x80d5201,0x80d3607,0x80d3601,0x80d3608,
-	0x80d3609,0x80d3604,0x80d3605,0x80d3606,0x80d360f,0x908c210,0xaa10008,0x680bfca,
-	0xc000ec0,0x4401208,0x908c210,0x7800309,0xb82b248,0x0000000,0x64d0009,0x9384249,
-	0xa6f8009,0x0000000,0x64d3609,0xa6ec009,0x0000000,0x920c290,0x7800349,0xb80c24a,
-	0x0000000,0x64d5309,0xaa1c009,0x0000000,0x60d2109,0x80d2001,0x64d2009,0xbefc049,
-	0x0000000,0xc7fe040,0x0000000,0x64d390f,0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,
-	0xc0006c0,0x7800308,0xc000e40,0x6800011,0x6a00014,0x6d00014,0x60d4d14,0xc001740,
-	0x680c00a,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,0x680c28a,0x401528a,
-	0x908c210,0x4002208,0x78002c9,0xb004248,0xc0009c0,0x908c450,0x680c50a,0xc001340,
-	0x0000000,0xc7ef200,0x0000000,0x7800289,0x240a248,0x540324a,0x540128a,0x2009289,
-	0x680c00a,0x200a289,0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x7400291,
-	0x400128a,0x5810451,0x7400291,0x400128a,0x7400292,0x400128a,0x5810492,0x7400292,
-	0x400128a,0x7400293,0x400128a,0x58104d3,0x7400293,0x400128a,0x7400294,0x400128a,
-	0x5810514,0xcc00000,0x7400294,0x690000e,0x6c0000e,0x200e30e,0x780028f,0x540434f,
-	0x54023cf,0x200f34f,0x200e3ce,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,0x603500d,
-	0x7c00291,0x4018251,0x7c002c9,0x5404251,0x5402211,0x2008248,0x690000e,0x6c0000e,
-	0x200e30e,0x200e20e,0x6803e8f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,
+	0xbefc108,0x0000000,0x80ac008,0x8007801,0xc7fd340,0x0000000,0xc010400,0x0000000,
+	0x6490b08,0x9c08608,0x9c08648,0x6090b08,0x6401108,0x9c08748,0x9c08788,0x6001108,
+	0x6401008,0x9808008,0x6001008,0x9c08008,0x6001008,0xc010840,0x0000000,0xc00ff40,
+	0x0000000,0xc07c7c0,0x0000000,0x64ac408,0x4001208,0x60ac408,0x64f4908,0x9c08048,
+	0x60f4908,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,0x6800008,0x6c00408,
+	0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a19408,0x60d1f08,0x64d1e08,0x98087c8,
+	0x60d1e08,0x64d3308,0x9008248,0x7c00089,0x9108248,0x7c000c9,0x9210208,0xc07c540,
+	0x60d2108,0xc07c4c0,0x80d2107,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,
+	0x7c00308,0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x6800008,0x7c00248,
+	0x60f1b08,0x7800309,0x946d209,0x60d3208,0x6940008,0x6c00008,0x2008308,0x60f2008,
+	0x6803908,0x64d3309,0x9008249,0x9508209,0x9808508,0x9808548,0x98087c8,0x60f1f08,
+	0x7800309,0x7800348,0x960f248,0x60f1e09,0xc07bc00,0x80d2100,0x7800148,0x7800189,
+	0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x78001c8,0xc07b980,0x60d2108,
+	0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,0x9605248,0x60d1f09,0x80d1f20,
+	0x80d1f20,0x64f1c08,0x9206208,0x7800049,0xc005580,0x2408248,0xc07b580,0x60d2108,
+	0x80d3d00,0x64acd08,0x60f4208,0xc0055c0,0x0000000,0x680d683,0x6813344,0x80d300f,
+	0x64d0008,0x9808008,0x9808048,0x9808088,0x9c080c8,0x9c08148,0x9c08188,0x9c081c8,
+	0x60d0008,0x64acf08,0x60f4808,0x6800289,0x78003c8,0x9485248,0x60f4009,0x80ac005,
+	0x8007801,0xc7fb000,0x0000000,0xc00eac0,0x0000000,0x64f4008,0x9c08048,0x60f4008,
+	0xc00df80,0x0000000,0x6490b08,0x9c08608,0x9808648,0x6090b08,0x6401108,0x9808748,
+	0x9c08788,0x6001108,0x6401008,0x9808008,0x6001008,0x9c08008,0x6001008,0xc00e3c0,
+	0x0000000,0xc00dac0,0x0000000,0xc07a340,0x0000000,0x64f4308,0x9c08048,0x9c080c8,
+	0x60f4308,0x64f1508,0x9c08508,0x9c084c8,0x9c08748,0x9c08448,0x9c08488,0x9c08348,
+	0x9c08308,0x9808248,0x9808008,0x60f1508,0x80d3400,0x9808488,0x9808448,0x9808348,
+	0x9808308,0x98085c8,0x9808608,0x9808648,0x9c08688,0x98086c8,0x9c08248,0x9c08008,
+	0x60f1508,0x64f4908,0x9c08048,0x9c081c8,0x9808008,0x60f4908,0x78003c9,0x5805249,
+	0x96e1209,0x5801249,0x9761209,0x98087c8,0x9808788,0x9808748,0x9808688,0x9808648,
+	0x9808608,0x9808288,0x9c08248,0x98081c8,0x9808088,0x9808048,0x60f4908,0x4801249,
+	0xaa10049,0x0000000,0xc781640,0x0000000,0x6800c08,0x6800809,0x94e8209,0x6800c09,
+	0x9608209,0x6801009,0x9708209,0x60f5608,0x6800008,0x6800009,0x9508209,0x60f5708,
+	0x680c0c8,0x6c00008,0x60f5808,0x68000c8,0x68000c9,0x9484209,0x68000c9,0x9504209,
+	0x6800049,0x9581209,0x6800009,0x95a1209,0x6800049,0x95c5209,0x60f5908,0x6800388,
+	0x6800489,0x9508209,0x6800149,0x9604209,0x68000c9,0x9684209,0x6800109,0x9704209,
+	0x6800089,0x9704209,0x60f5a08,0x6800008,0x6800009,0x9508209,0x6800809,0x9608209,
+	0x6801409,0x9708209,0x60f5b08,0x6801008,0x6800009,0x9508209,0x6801809,0x9608209,
+	0x6803009,0x9708209,0x60f5c08,0x6800608,0x6801409,0x9508209,0x6800809,0x9608209,
+	0x6801809,0x9708209,0x60f5d08,0x68000c8,0x6801809,0x948c209,0x6800249,0x9607209,
+	0x6800009,0x96e1209,0x6800c09,0x9708209,0x60f5e08,0x68000c8,0x6800009,0x9461209,
+	0x6800049,0x9481209,0x6800009,0x94a2209,0x6803009,0x950c209,0x6801c09,0x968c209,
+	0x60f5f08,0xc7807c0,0x0000000,0x6804008,0x68000c9,0x9586209,0x6800089,0x9646209,
+	0x9808608,0x9808648,0x9808688,0x98086c8,0x9808708,0x9808748,0x9808788,0x98087c8,
+	0x60f5508,0x6800c08,0x6800809,0x94e8209,0x6800c09,0x9608209,0x6801009,0x9708209,
+	0x60f5608,0x6800c08,0x6801009,0x9508209,0x60f5708,0x680c0c8,0x6c0c0c8,0x60f5808,
+	0x64acc09,0x60f4a09,0x64d3109,0x9010209,0x400f208,0x5804208,0x4401208,0x7c00308,
+	0x9210209,0x400f208,0x5804208,0x4401208,0x7c00348,0x7800309,0x7800348,0x958c248,
+	0x6499008,0x9708248,0x60f4f09,0x6bc0008,0x60d1f08,0x64d1e08,0x9c087c8,0x60d1e08,
+	0x6800008,0x6c00408,0x60d1f08,0x6800048,0x6c00408,0x60d1f08,0x6a10408,0x60d1f08,
+	0x64d1e08,0x98087c8,0x60d1e08,0x64d3308,0x9210208,0xc0773c0,0x60d2108,0xc077340,
+	0x80d2105,0x6800048,0x7c00248,0x60f1b08,0x7800309,0x946d209,0x60d3208,0x7800348,
+	0x960f248,0x60f1e09,0xc077080,0x80d2100,0x7800148,0x7800189,0x680040a,0x240a20a,
+	0x3409289,0x9605248,0x60d1f09,0x7800108,0x7800209,0x680040a,0x240a20a,0x3409289,
+	0x9605248,0x60d1f09,0x80d1f20,0x80d1f20,0x80d1f20,0x64f1d08,0x9206208,0x7800049,
+	0xc000a80,0x2408248,0xc076a80,0x60d2108,0x80d3d00,0x80d5300,0x64acd08,0x60f4208,
+	0xc001840,0x0000000,0x6810143,0x6813bc4,0x6800008,0x6c00448,0x60d3008,0x64d0008,
+	0x9c08008,0x9c08048,0x9c08088,0x9c08188,0x9c081c8,0x60d0008,0x9808008,0x9808048,
+	0x9808088,0x98080c8,0x9808148,0x9808188,0x98081c8,0x60d0008,0x6800048,0x7800309,
+	0x948c209,0x60f4b08,0x80ac005,0x8007801,0xc7f6340,0x0000000,0x0800000,0x7800008,
+	0xc7f5a40,0x6030408,0x6800009,0xb005248,0x2409209,0x5401208,0xcc00000,0x4401208,
+	0xcc00000,0x5401209,0x6bfffda,0x6c0001a,0x6800009,0x680001b,0x6c0001b,0x400071b,
+	0x680000b,0x680c00e,0x680100f,0x740039a,0x400138e,0x7400389,0x400138e,0x740039b,
+	0x400138e,0x740038b,0x400138e,0xaee004f,0x44013cf,0x690000e,0x6c0000e,0x200e30e,
+	0x680000f,0x603510e,0x680400d,0x603520d,0x6a0c00d,0x603500d,0x643500d,0x580f34d,
+	0xbef804d,0x0000000,0x40403cf,0xb611e8f,0x680800d,0xc7ffd00,0x200e34e,0x64d330d,
+	0x900834d,0x680000e,0x680100f,0xb8053cd,0x0000000,0x404038e,0xc7fff40,0x40403cf,
+	0x7c0028e,0x44013cf,0x7c002cf,0x60d371a,0x60d391b,0x60d381a,0x60d3a1b,0xcc00000,
+	0x0000000,0x8098002,0x8098000,0x6a0001d,0x6d0001d,0x6a0001e,0x6d0001e,0x6a0001f,
+	0x6d0001f,0x6a00020,0x6d00020,0x690000a,0x680001b,0x6c0001b,0x400071b,0x6800009,
+	0x680c00e,0x680064f,0x740039b,0x400138e,0x7400389,0x400138e,0x740039d,0x400138e,
+	0x740038a,0x400138e,0x740039e,0x400138e,0x740038a,0x400138e,0x740039f,0x400138e,
+	0x740038a,0x400138e,0x74003a0,0x400138e,0x740038a,0x400138e,0xaeb004f,0x44013cf,
+	0x690000e,0x6c0000e,0x200e30e,0x680000f,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,
+	0x603500d,0x643500d,0x580f34d,0xbef804d,0x0000000,0x40193cf,0xb611e8f,0x6807d0d,
+	0xc7ffd00,0x200e34e,0x64d330d,0x900834d,0x680000e,0x680064f,0xb8053cd,0x0000000,
+	0x401938e,0xc7fff40,0x40193cf,0x7c0028e,0x44013cf,0x7c002cf,0x60d391b,0x60d4a1d,
+	0x60d4b1e,0x60d4c1f,0x60d4d20,0x60d4a1d,0x60d4a1d,0x60d3a1b,0x60d4e1d,0x60d4f1e,
+	0x60d501f,0x60d5120,0xcc00000,0x0000000,0xaa10008,0x64d370e,0xcb80003,0x64d390f,
+	0xc7f3440,0x0000000,0x60d371a,0x64d3510,0x908c250,0xa620009,0x60d391b,0x6bfffc9,
+	0x6c00009,0x60d3809,0x6800009,0x6c00009,0x60d3a09,0x9004250,0xaa1c249,0x80d3601,
+	0x80d3603,0x80d3605,0x80d3606,0xc780200,0x80d360f,0x80d3602,0x80d3603,0x80d3604,
+	0x80d3605,0x80d3606,0x80d360f,0x908c210,0x680c10a,0x7800289,0xa814248,0x2409248,
+	0x4401249,0x5402249,0x680c00a,0x200a289,0x740028e,0x400128a,0x581038e,0x740028e,
+	0x400128a,0x740028f,0x400128a,0x58103cf,0x740028f,0x400528a,0x78002c9,0xb40d248,
+	0x7800309,0xb40b248,0x700029a,0x400128a,0x700028e,0x961068e,0x400128a,0x700029b,
+	0x400128a,0x700028f,0xc7f25c0,0x96106cf,0x908c210,0x7800289,0x2409248,0x5402249,
+	0x680c00a,0x200a289,0x64d0009,0x9384249,0xa6f8009,0x0000000,0x64d3609,0xa6ec009,
+	0x0000000,0x64d370e,0x64d390f,0x740028e,0x400128a,0x581038e,0x740028e,0x400128a,
+	0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x690000e,0x6c0000e,0x200e30e,
+	0x780028f,0x54033cf,0x200e3ce,0x603510e,0x680400d,0x603520d,0x6a0c00d,0x603500d,
+	0x908c210,0x7800309,0xb830248,0x4001208,0x920c210,0x7800349,0xb82c248,0x6800008,
+	0x80d3000,0x80d1f30,0x6bc0008,0x60d1f08,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x0000000,0x0000000,0x64d1608,0xbefc108,0x0000000,0x6800008,0x6e00008,0x60d1c08,
+	0x80d1d00,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x64d1608,
+	0xbefc108,0x0000000,0x64d1b08,0x9808648,0x60d1b08,0x64d1608,0xbefc108,0x0000000,
+	0x7800188,0x4001208,0x7c00188,0x7800208,0x4002208,0x7c00208,0x80ac009,0x8007801,
+	0xc7f1040,0x0000000,0x7c00288,0x403f248,0x7c002c9,0x5403208,0x690000e,0x6c0000e,
+	0x200e30e,0x200e20e,0x680400f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,
 	0x643530d,0x900c34d,0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,0x643500d,
-	0x580f34d,0xbef804d,0xcc00000,0x0000000,0x40006dc,0x700029c,0x400128a,0x7000289,
-	0x9610709,0x400128a,0x700029d,0x400128a,0x7000289,0x9610749,0x400128a,0x700029e,
-	0x400128a,0x7000289,0x9610789,0x400128a,0x700029f,0x400128a,0x7000289,0x96107c9,
-	0x400128a,0x70002a0,0x400128a,0x7000289,0xcc00000,0x9610809,0x64d1602,0x9162082,
-	0xaa0c002,0xc7fff40,0x0000000,0xc7ed780,0x0000000,0xcb80004,0x64f1f08,0x9361248,
-	0xaa10009,0x9042248,0xc7ed5c0,0x0000000,0xaa14009,0x680324a,0xaa0c049,0x680348a,
-	0x680390a,0x7800088,0x7800309,0xb806248,0x4001208,0x6800008,0x78000c9,0x4001249,
-	0x7c000c9,0x7c00088,0x950c288,0xa60c008,0x7800309,0x980a50a,0xa403248,0x0000000,
-	0x980a58a,0x78000c8,0xa60c008,0x60f4108,0x980a54a,0x980a7ca,0x60f1f0a,0xc7ece80,
-	0x0000000,0x64f4b08,0x9384248,0xaa10009,0x64f4a08,0xc7ecd00,0x0000000,0x920c248,
-	0x780034a,0xb408289,0x4001249,0x960c209,0x60f4a08,0x6800048,0x7800309,0x948c209,
-	0x60f4b08,0xc7eca00,0x0000000,0x64ac10c,0x680c008,0x6094008,0x6803009,0x6804008,
-	0xd000009,0x6094108,0x6800008,0x6094008,0xcc00000,0x0000000,0xcc00000,0x0000000,
-	0xc000a00,0x0000000,0x780034a,0x400128a,0x40002ca,0x7800309,0x4001249,0x96102c9,
-	0x609520b,0x54042ca,0x968c2c9,0x609080b,0x681010a,0x6c3454a,0x609070a,0x6490b0a,
-	0x980a00a,0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,0x698000a,0x6c0000a,
-	0x200a30a,0x609440a,0x6a0000a,0x6c0000a,0x200a30a,0x609450a,0xcc00000,0x0000000,
-	0xc0005c0,0x0000000,0x8095003,0x8095000,0x6820889,0x6c00209,0xcc00000,0x6095109,
-	0x683ffc9,0x649090a,0xa21004a,0xaa0c009,0xc7fff40,0x4401249,0x6490b0a,0x980a00a,
-	0x980a0ca,0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,0xcc00000,0x0000000,0x683ffca,
-	0x649530b,0x90012cb,0x649bb09,0x920c249,0x20092c9,0xaa10009,0xaa0c00a,0xc7ffe40,
-	0x440128a,0xcc00000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
+	0x580f34d,0xbef804d,0x0000000,0xc7fe1c0,0x680c00a,0x64d4a11,0x64d4b12,0x64d4c13,
+	0x64d4d14,0x60d4a1d,0x60d4b1e,0x60d4c1f,0x60d4d20,0x64d3510,0x908c250,0xa62c009,
+	0x60d391b,0x6a00009,0x6d00009,0x60d4e09,0x60d4f09,0x60d5009,0x60d5109,0x6800009,
+	0x6c00009,0x60d3a09,0x80d5201,0x80d3607,0x80d3601,0x80d3608,0x80d3609,0x80d3604,
+	0x80d3605,0x80d3606,0x80d360f,0x908c210,0xaa10008,0x680bfca,0xc000ec0,0x4401208,
+	0x908c210,0x7800309,0xb82b248,0x0000000,0x64d0009,0x9384249,0xa6f8009,0x0000000,
+	0x64d3609,0xa6ec009,0x0000000,0x920c290,0x7800349,0xb80c24a,0x0000000,0x64d5309,
+	0xaa1c009,0x0000000,0x60d2109,0x80d2001,0x64d2009,0xbefc049,0x0000000,0xc7fe040,
+	0x0000000,0x64d390f,0x64d4a11,0x64d4b12,0x64d4c13,0x64d4d14,0xc0006c0,0x7800308,
+	0xc000e40,0x6800011,0x6a00014,0x6d00014,0x60d4d14,0xc001740,0x680c00a,0x60d4a1d,
+	0x60d4b1e,0x60d4c1f,0x60d4d20,0xc780280,0x680c28a,0x401528a,0x908c210,0x4002208,
+	0x78002c9,0xb004248,0xc0009c0,0x908c450,0x680c50a,0xc001340,0x0000000,0xc7ef280,
+	0x0000000,0x7800289,0x240a248,0x540324a,0x540128a,0x2009289,0x680c00a,0x200a289,
+	0x740028f,0x400128a,0x58103cf,0x740028f,0x400128a,0x7400291,0x400128a,0x5810451,
+	0x7400291,0x400128a,0x7400292,0x400128a,0x5810492,0x7400292,0x400128a,0x7400293,
+	0x400128a,0x58104d3,0x7400293,0x400128a,0x7400294,0x400128a,0x5810514,0xcc00000,
+	0x7400294,0x690000e,0x6c0000e,0x200e30e,0x780028f,0x540434f,0x54023cf,0x200f34f,
+	0x200e3ce,0x603510e,0x6803e8d,0x603520d,0x6a0c00d,0x603500d,0x7c00291,0x4018251,
+	0x7c002c9,0x5404251,0x5402211,0x2008248,0x690000e,0x6c0000e,0x200e30e,0x200e20e,
+	0x6803e8f,0x6b0c010,0x643500d,0x580f34d,0xbef804d,0x0000000,0x643530d,0x900c34d,
+	0xa6f800d,0x0000000,0x603510e,0x603520f,0x6035010,0x643500d,0x580f34d,0xbef804d,
+	0xcc00000,0x0000000,0x40006dc,0x700029c,0x400128a,0x7000289,0x9610709,0x400128a,
+	0x700029d,0x400128a,0x7000289,0x9610749,0x400128a,0x700029e,0x400128a,0x7000289,
+	0x9610789,0x400128a,0x700029f,0x400128a,0x7000289,0x96107c9,0x400128a,0x70002a0,
+	0x400128a,0x7000289,0xcc00000,0x9610809,0x64d1602,0x9162082,0xaa0c002,0xc7fff40,
+	0x0000000,0xc7ed800,0x0000000,0xcb80004,0x64f1f08,0x9361248,0xaa10009,0x9042248,
+	0xc7ed640,0x0000000,0xaa14009,0x680324a,0xaa0c049,0x680348a,0x680390a,0x7800088,
+	0x7800309,0xb806248,0x4001208,0x6800008,0x78000c9,0x4001249,0x7c000c9,0x7c00088,
+	0x950c288,0xa60c008,0x7800309,0x980a50a,0xa403248,0x0000000,0x980a58a,0x78000c8,
+	0xa60c008,0x60f4108,0x980a54a,0x980a7ca,0x60f1f0a,0xc7ecf00,0x0000000,0x64f4b08,
+	0x9384248,0xaa10009,0x64f4a08,0xc7ecd80,0x0000000,0x920c248,0x780034a,0xb408289,
+	0x4001249,0x960c209,0x60f4a08,0x6800048,0x7800309,0x948c209,0x60f4b08,0xc7eca80,
+	0x0000000,0x64ac10c,0x680c008,0x6094008,0x6803009,0x6804008,0xd000009,0x6094108,
+	0x6800008,0x6094008,0xcc00000,0x0000000,0xcc00000,0x0000000,0xc000a00,0x0000000,
+	0x780034a,0x400128a,0x40002ca,0x7800309,0x4001249,0x96102c9,0x609520b,0x54042ca,
+	0x968c2c9,0x609080b,0x681010a,0x6c3454a,0x609070a,0x6490b0a,0x980a00a,0x980a0ca,
+	0x6090b0a,0x9c0a00a,0x9c0a0ca,0x6090b0a,0x698000a,0x6c0000a,0x200a30a,0x609440a,
+	0x6a0000a,0x6c0000a,0x200a30a,0x609450a,0xcc00000,0x0000000,0xc0005c0,0x0000000,
+	0x8095003,0x8095000,0x6820889,0x6c00209,0xcc00000,0x6095109,0x683ffc9,0x649090a,
+	0xa21004a,0xaa0c009,0xc7fff40,0x4401249,0x6490b0a,0x980a00a,0x980a0ca,0x6090b0a,
+	0x9c0a00a,0x9c0a0ca,0x6090b0a,0xcc00000,0x0000000,0x683ffca,0x649530b,0x90012cb,
+	0x649bb09,0x920c249,0x20092c9,0xaa10009,0xaa0c00a,0xc7ffe40,0x440128a,0xcc00000,
+	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000
 };
 
 
diff --git a/drivers/amlogic/amports/vdec.c b/drivers/amlogic/amports/vdec.c
index dffe0ec635ad..0271e532f471 100755
--- a/drivers/amlogic/amports/vdec.c
+++ b/drivers/amlogic/amports/vdec.c
@@ -562,41 +562,6 @@ static struct class vdec_class = {
         .class_attrs = vdec_class_attrs,
     };
 
-#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON8
-const u32 fix_mc[] __attribute__ ((aligned (8))) = {
-    0x0809c05a, 0x06696000, 0x0c780000, 0x00000000
-};
-
-void vdec_dos_top_reg_fix(void)
-{
-    bool hcodec_on = vdec_on(VDEC_HCODEC);
-
-    if ((hcodec_on) && (READ_VREG(HCODEC_MPSR) & 1)) {
-        return;
-    }
-
-    /* this function should be called only once after power on
-     * to work around a HW bug for dual video decoders access
-     * DOS top level CBUS registers
-     */
-    if (!hcodec_on) {
-        vdec_poweron(VDEC_HCODEC);
-    }
-
-    amhcodec_loadmc(fix_mc);
-
-    amhcodec_start();
-
-    udelay(1000);
-
-    amhcodec_stop();
-
-    if (!hcodec_on) {
-        vdec_poweroff(VDEC_HCODEC);
-    }
-}
-#endif
-
 s32 vdec_dev_register(void)
 {
     s32 r;
@@ -607,10 +572,6 @@ s32 vdec_dev_register(void)
         return r;
     }
 
-#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON8
-    vdec_dos_top_reg_fix();
-#endif
-
 #if MESON_CPU_TYPE < MESON_CPU_TYPE_MESON8
     /* default to 250MHz */
     vdec_clock_hi_enable();
diff --git a/drivers/amlogic/amports/vdec.h b/drivers/amlogic/amports/vdec.h
index a8bc6684d206..9435c219819c 100755
--- a/drivers/amlogic/amports/vdec.h
+++ b/drivers/amlogic/amports/vdec.h
@@ -44,7 +44,6 @@ extern void vdec2_power_mode(int level);
 extern void vdec_poweron(vdec_type_t core);
 extern void vdec_poweroff(vdec_type_t core);
 extern bool vdec_on(vdec_type_t core);
-extern void vdec_dos_top_reg_fix(void);
 #else
 #define vdec_poweron(core)
 #define vdec_poweroff(core)
diff --git a/drivers/amlogic/amports/vh264_4k2k.c b/drivers/amlogic/amports/vh264_4k2k.c
index f461d986bf28..3957363b4b47 100755
--- a/drivers/amlogic/amports/vh264_4k2k.c
+++ b/drivers/amlogic/amports/vh264_4k2k.c
@@ -55,6 +55,8 @@
 #define STAT_TIMER_ARM      0x10
 #define STAT_VDEC_RUN       0x20
 
+extern void amvenc_dos_top_reg_fix(void);
+
 static int  vh264_4k2k_vf_states(vframe_states_t *states, void*);
 static vframe_t *vh264_4k2k_vf_peek(void*);
 static vframe_t *vh264_4k2k_vf_get(void*);
@@ -1111,6 +1113,8 @@ static void vh264_4k2k_prot_init(void)
 
     WRITE_VREG(MDEC_PIC_DC_THRESH, 0x404038aa);
     WRITE_VREG(VDEC2_MDEC_PIC_DC_THRESH, 0x404038aa);
+
+    amvenc_dos_top_reg_fix();
 }
 
 static void vh264_4k2k_local_init(void)
diff --git a/drivers/amlogic/amports/vreal.c b/drivers/amlogic/amports/vreal.c
index 510a10a6c4a8..1c049289866c 100755
--- a/drivers/amlogic/amports/vreal.c
+++ b/drivers/amlogic/amports/vreal.c
@@ -860,10 +860,6 @@ static int amvdec_real_remove(struct platform_device *pdev)
 
     amvdec_disable();
 
-#if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON8
-    vdec_dos_top_reg_fix();
-#endif
-
     printk("frame duration %d, frames %d\n", frame_dur, frame_count);
     return 0;
 }
-- 
2.19.0

